### PR TITLE
[search] Ranker takes into account possible errors in the street part.

### PR DESCRIPTION
**Pull request by ygorshenin**
@mpimenov @bykoianko PTAL

### CODE REVIEW

**Comment by ygorshenin** [comment]
Dear colleagues, please, comment WDYT about combined name-score.

File: search/geocoder.hpp [hunk 1]
```diff
@@ -196,7 +196,8 @@ class Geocoder
 
   // Forms result and feeds it to |m_preRanker|.
   void EmitResult(BaseContext const & ctx, MwmSet::MwmId const & mwmId, uint32_t ftId,
-                  SearchModel::SearchType type, TokenRange const & tokenRange);
+                  SearchModel::SearchType type, TokenRange const & tokenRange,
+                  IntersectionResult const * m_geoParts);
```
Comment (ygorshenin) [inline search/geocoder.hpp]: Done

File: search/pre_ranking_info.hpp [hunk 2]
```diff
@@ -39,6 +40,9 @@ struct PreRankingInfo
   // Tokens match to the feature name or house number.
   TokenRange m_tokenRange[SearchModel::SEARCH_TYPE_COUNT];
 
+  // Different geo-parts participated in search.
```
Comment (ygorshenin) [inline search/pre_ranking_info.hpp]: Done

File: search/intersection_result.hpp [hunk 3]
```diff
@@ -2,18 +2,19 @@
...
 {
 // This class holds higher-level features for an intersection result,
 // i.e. BUILDING and STREET for POI or STREET for BUILDING.
 struct IntersectionResult
 {
-  static uint32_t const kInvalidId;
+  static uint32_t const kInvalidId = std::numeric_limits<uint32_t>::max();
 
-  IntersectionResult();
+  IntersectionResult() = default;
```
Comment (bykoianko) [inline search/intersection_result.hpp]: Is it possible to remove this line?

Comment (ygorshenin) [inline search/intersection_result.hpp]: Done

**Comment by ygorshenin** [comment]
@mpimenov @bykoianko PTAL

File: search/intersection_result.hpp [hunk 4]
```diff
@@ -2,18 +2,17 @@
...
 namespace search
 {
 // This class holds higher-level features for an intersection result,
 // i.e. BUILDING and STREET for POI or STREET for BUILDING.
 struct IntersectionResult
 {
-  static uint32_t const kInvalidId;
-
-  IntersectionResult();
+  static uint32_t const kInvalidId = std::numeric_limits<uint32_t>::max();
```
Comment (bykoianko) [inline search/intersection_result.hpp]: constexpr?

File: search/search_integration_tests/ranker_test.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
+#include "testing/testing.hpp"
+
+#include "search/search_integration_tests/helpers.hpp"
+#include "search/search_tests_support/test_results_matching.hpp"
+
+#include "generator/generator_tests_support/test_feature.hpp"
+
+
```
Comment (bykoianko) [inline search/search_integration_tests/ranker_test.cpp]: Let's remove the line.

File: search/intersection_result.hpp (hunk 4 above)
Comment (ygorshenin) [inline search/intersection_result.hpp]: Okay, but I can't see any practical benefit from this.

File: search/search_integration_tests/ranker_test.cpp (hunk 5 above)
Comment (ygorshenin) [inline search/search_integration_tests/ranker_test.cpp]: Done

File: search/ranker.cpp [hunk 6]
```diff
@@ -141,29 +166,36 @@ class PreResult2Maker
...
+    if (!m_loader->GetFeatureByIndex(id.m_index, ft))
       return false;
 
-    f.SetID(id);
+    ft.SetID(id);
+    return true;
+  }
 
-    center = feature::GetCenter(f);
+  // For the best performance, incoming id's should be sorted by id.first (mwm file id).
```
Comment (mpimenov) [inline search/ranker.cpp]: incomiing ids

Comment (ygorshenin) [inline search/ranker.cpp]: Done

**Comment by ygorshenin** [comment]
PTAL

**Comment by mpimenov** [comment]
LGTM

**Pull request merged**